### PR TITLE
Additional SHACL support to Ontospy/gendocs

### PR DESCRIPTION
**Pull request by balon**
This patch series is intended to bring Ontospy's current SHACL support up to the level of displaying Property Shapes
on class pages and Shape Properties on shape pages.

A new module, _ontospy/core/shacl_helper.py_, returns SHACL property constraints, is attached to each OntoClass object, 
once the shapes are built in Ontospy.

**A few notes on changes:**
- The modifications made in 4a6d48c method of linking nodes in tree-diagram were done to ensure the diagram was still properly linking to other shapes (this is achieved by using a django filter and 'forwarding' the slugify function from core/utils)
- The modifications made in 1750e87 will only display if an ontology uses SHACL

Please advise if there is a better approach to increasing the support for SHACL into Ontospy, or if any modifications are 
needed to the current patch presented.

### CODE REVIEW

**Comment by lambdamusic** [comment]
Thanks. I'll be able to review it over the next few days. 
If you could attach/reference a sample rdf file for testing all new features that'd be great.

**Comment by balon** [comment]
Attached is uco_monolithic.ttl, which can be used for testing the features in this patch series. The file _could_ also be constructed by running the makefile found in the [ucoProject/UCO](https://github.com/ucoProject/UCO/blob/master/tests/Makefile) repository.

Thanks!

[uco_monolithic.ttl.zip](https://github.com/lambdamusic/Ontospy/files/7694568/uco_monolithic.ttl.zip)

**Comment by ajnelson-nist** [comment]
This PR has received an in-flight update, for the same reason noted over on [PR 108](https://github.com/lambdamusic/Ontospy/pull/108#issuecomment-995947103).

**Comment by tobiasschweizer** [comment]
Hi

I added a feature request in #104 and I think you PR does what I strive for :-)

I did the following:
- I cloned https://github.com/casework/Ontospy/tree/feature-SHACL-gendocs
- I ran `python setup.py install` and installed `pygments` and `django<4`(#112)
- I ran `ontospy gendocs uco_monolithic.ttl` from my venv

What I get as HTML looks correct in terms of property shapes on the class page. This is really nice!

However, I tried this with my own ontology which consists of
- an ontology part for schema.org classes
- a SHACL part that targets schema.org classes via `sh:targetClass`

As observed in https://github.com/lambdamusic/Ontospy/pull/107#issuecomment-997284164 I don't see the Shapes page anymore.

Could this be because classes and node shapes do not use the same IRI? For some more details about the ontology I use, please see #104 

Thanks a lot!

**Comment by ajnelson-nist** [comment]
Thank you both for the feedback!  We'll be getting back to you next week.

**Comment by balon** [comment]
Thank you for the feedback. We had a discussion today about a timeline to updating the PR as requested. We're both currently tied up on a few tasks, but plan to address this early next week with plans to have something ready to push by next Friday.

We will also look at the issues presented involving the missing shapes page on some ontologies.

**Comment by lambdamusic** [comment]
Sounds good, thanks. If there are any updates at my end, I'll let you know.

**Comment by ajnelson-nist** [comment]
We're working through the remarks now.  We've made a catch-up merge to integrate `master`'s current state, and have converted the branch to a draft state while we step through the other remarks.

@lambdamusic , re:
> For some reason the documentation generation is backward incompatible. See for example this documentation page for SciGraph - RDF source here - previously I was able to generate a 'Shapes' page for SciGraph, but with your patch I get none. Any idea why?

Could you explain what behavior you're seeing here?  After our catch-up merge, we regenerated the page, and saw these effects:
* `https://lambdamusic.github.io/ontospy-examples/scigraph-core-ontologyttl/index.html` - this page's "Metrics" table lists 17 data shapes, which matches the current deployed state on ontospy-examples.
* `https://lambdamusic.github.io/ontospy-examples/scigraph-core-ontologyttl/entities-tree-shapes.html` - this page lists 17 shapes.
* `https://lambdamusic.github.io/ontospy-examples/scigraph-core-ontologyttl/shape-shapesaffiliation.html` - this page does not list a properties table.  However, we believe that is because we search for properties based on `sh:path` being defined in the `sh:PropertyShape`.  (Our added queries do not search for an explicit class declaration of `sh:PropertyShape`; we instead assume that based on `sh:property` linking the blank node and implying it's a property shape.)

I suspect we've actually resolved your second issue, but that's mainly from our failure to reproduce what you saw.

**Comment by ajnelson-nist** [comment]
On further review, `sh:predicate` is not defined in SHACL.  Did the scigraph RDF source intend to use [`sh:path`](https://www.w3.org/TR/shacl/#property-shapes)?

**Comment by ajnelson-nist** [comment]
On the shapes page, we intend to list a table of how properties associate with a class.  There's a conceptual hop there, which is probably something we need to talk through to come to a good design consensus.

The user interface complexity we're trying to hide is going to apply to SHACL-based schemes that don't define `owl:Restriction`s.  In our use case, we don't have `owl:Restriction`s, but we do have `sh:PropertyShape`s.  So, our property associations would be listed on the Shapes HTML page.  But, to see things about other classes in the parent/child hierarchy, the user would hop over to the Class page, where no properties would be listed.

So, we tried to apply some redundancy - having the properties associated via SHACL be listed on the class page as well.  This saves the user **two clicks** (oscillating between shapes and class pages) per level of the class hierarchy they want to review.

We probably wouldn't go for this redundancy if SHACL had a concept of "SubNodeShape," but they don't.

That's the feedback we could get to today, we're happy to discuss this further.

**Comment by lambdamusic** [comment]
Hi folks - first of all, wish I had time to get back to you sooner, sorry about that and thanks a lot for the patience. 
I reviewed the PO after the merging, have three main comments.  

### 1) Shapes HTML page for scigraph is ok
@ajnelson-nist as you pointed out, the scigraph ontology 'shapes' page is not missing anymore. Don't know what's changed, I do recall noticing the difference, but of course I might be wrong. So all's good there. Backward compatibility is fine. 

### 2) New ontospy tests error 
I do get an error when running the test suite (`./tools/run-all-tests.sh`). The error is triggered when trying to scan the [schema](https://github.com/lambdamusic/Ontospy/tree/master/ontospy/tests/rdf/schema) test RDF model. You can reproduce simply by typing `$ ontospy scan ontospy/tests/rdf/schema/` (on this branch, of course). Here's the traceback: 

```bash
...
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/cli.py", line 181, in scan
    action_analyze(sources, endpoint, print_opts, verbose, extra, raw, individuals, format)
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/actions.py", line 123, in action_analyze
    g = Ontospy(
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/ontospy.py", line 134, in __init__
    self.build_all(
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/ontospy.py", line 249, in build_all
    self.build_shapes()
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/ontospy.py", line 635, in build_shapes
    shacl_constraints = build_shacl_constraints(self)
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/shacl_helper.py", line 255, in build_shacl_constraints
    add_property_constraints_from_shape_triples(lineage_class.uri, property_constraints, shape.triples)
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/shacl_helper.py", line 298, in add_property_constraints_from_shape_triples
    raise Exception('Something is very wrong.  Class URI not in triples.')
Exception: Something is very wrong.  Class URI not in triples.
```

The error is triggered by line [line 297](https://github.com/lambdamusic/Ontospy/blob/casework-feature-SHACL-gendocs/ontospy/core/shacl_helper.py#L297). Not really sure I fully understand the logic there, so maybe you can shed some light. 

Depending on how bad the cause of the error is, I'd suggest issuing a warning instead and/or failing silently. 
Ontospy is not really a schema validator, so I'd prefer to render an incomplete output rather than failing completely. 


### 3) SciGraph shapes error

You are totally right @ajnelson-nist, `sh:predicate` should be [sh:path](https://www.w3.org/TR/shacl/#property-shapes). Not sure what happened there, but I suppose when we built that model a few years back we may have used an older Shacl specification which is [now deprecated](https://w3c.github.io/data-shapes/unmaintained/shacl-abstract-syntax/). Who knows.. 

Anyhow, good that you spotted it. I've updated the test RDF with `sh:path` within a new folder `scigraph2` and pushed it [to this branch](https://github.com/lambdamusic/Ontospy/tree/casework-feature-SHACL-gendocs/ontospy/tests/rdf/scigraph2). Tried to run ontospy on it, but also here I'm getting the error above: 

```bash
  File "/Users/michele.pasin/code/python/ontospy/ontospyProject/ontospy/core/shacl_helper.py", line 298, in add_property_constraints_from_shape_triples
    raise Exception('Something is very wrong.  Class URI not in triples.')
Exception: Something is very wrong.  Class URI not in triples.
```

So, even more curious about how to get rid of that. 

---

**To sum up** - I believe that after addressing that validation error above, we're good to go :-) 

The UI issues you mentioned are important, but maybe better tackled outside this PR and by using more test input data.

**Comment by ajnelson-nist** [comment]
First, @tobiasschweizer , you were right to give a poke, thanks for that.

@lambdamusic , thank you for the test case.  We stumbled on a similar issue.  @balon and I have started a type review of the `shacl_helper.py` script on observing that there were a few errant assumptions about nodes being required to be blank nodes or `rdflib.URIRef`s.

**Comment by ajnelson-nist** [comment]
@lambdamusic : @balon and I think the "Something is very wrong" exception has been addressed.  I think we've covered the tests you've mentioned above; `./tools/run-all-tests.sh` and `nosetests` both pass.  If there's nothing else, this might be good to merge.

**Comment by lambdamusic** [comment]
Available in https://github.com/lambdamusic/Ontospy/releases/tag/2.0.0-alpha

**Comment by ajnelson-nist** [comment]
Thank you!

**Comment by tobiasschweizer** [comment]
Hi there,

That is great news, I'll try this asap.
